### ISSUE TITLE
[JENKINS-3265] Reload Configuration from Disk (or POSTing config.xml) loses info on running builds

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-06-a2#2350]
<p>Executing the "Reload Configuration from Disk" while builds are running causes<br/>
information about those builds to be lost.</p>

<p>POSTing a new configuration to /job/JOBNAME/config.xml also causes the exact same problem. Yet editing job configuration interactively using the web form does not show the problem.</p>

<p>I had two builds running when I reloaded configuration to pick up manual changes<br/>
to one project (not running a build) config.xml file.  When the browser<br/>
refreshed, the builds were still shown as actively building in the Build<br/>
Executor Status frame but the status indicators were no longer blinking and the<br/>
project pages did not show any builds running for those projects.</p>

<p>It appears that part of Hudson thought that the builds had been canceled because<br/>
they were queued, blocked by themselves.  The original builds appeared to<br/>
finsish normally and then the queued builds ran.  There was no record of the<br/>
original build that had been running when the config reloaded.</p>

<p>Note: If Jenkins is completely restarted, by shutting down and starting up the service, the missing build re-appears. (Of course, doing that each time is unacceptable, just trying to help focus the debugging.)</p>

---
<details><summary><i>Originally reported by <a title="pcampbell" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=pcampbell">pcampbell</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-3265" target="_blank">Reload Configuration from Disk (or POSTing config.xml) loses info on running builds</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Kohsuke Kawaguchi" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>
<li><b>reported by</b>: pcampbell
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2015-02-11T12:40:15+00:00
<li><b>votes</b>: 28
<li><b>watchers</b>: 49
<li><b>imported</b>: 2025-11-06
</ul></i>
</details>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=shurikgefter">shurikgefter</a>:</i>
<p>I have the same problem.</p>

<p>I checked it with Hudson #3.369 and #3.383</p>

<p>How reproduce the problem:</p>

<p>1)	Define the free-style project.<br/>
2)	Add command sleep 300<br/>
3)	Run this job.<br/>
4)	When this job running execute "Reload Configuration from Disk"<br/>
5)	Click on progress bar of running job ( console output )<br/>
6)	Get page 404 error<br/>
7)	In the history of this job I can't see the current running execution.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=samweiss">samweiss</a>:</i>
<p>Hi</p>

<p>I too have this problem, and the way I manage my builds, this is getting very troublesome (I have automatic processes to create jobs for me for build branches that are created).</p>

<p>Any time Hudson reloads configuratons, not only are the info on running builds lost, you have to do another reload configuration once a running job is completed in order for hudson to recognize it.</p>

<p>Thanks!</p>

<p>Sam</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=asull">asull</a>:</i>
<p>We are also struggling with this problem: it makes it problematic to use the reload from disk option because we have a lot of long-running (>1hr) builds, and we don't want to lose visibility of currently executing jobs.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=samweiss">samweiss</a>:</i>
<p>Is there a status on this bug?  It was filed over a year ago, and still repros.  It's listed as "major"!  Hello????</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=pmv">pmv</a>:</i>
<p>David - we also switched to using the cli to perform bulk changes.  I currently use the following groovy method to make a job isn't running before doing the update:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"><span class="code-keyword">public</span> <span class="code-object">boolean</span> isJobRunning(<span class="code-object">String</span> jobName) {
	URL jobDetails = <span class="code-keyword">new</span> URL(JENKINS_URL + <span class="code-quote">"/job/"</span> + jobName + <span class="code-quote">"/api/xml?depth=1"</span>)
	SAXParserFactory factory = SAXParserFactory.newInstance()
	factory.setFeature( <span class="code-quote">"http:<span class="code-comment">//apache.org/xml/features/nonvalidating/load-external-dtd"</span>, <span class="code-keyword">false</span>)
</span>	def details = <span class="code-keyword">new</span> XmlSlurper(factory.newSAXParser()).parse(jobDetails.openStream())
	<span class="code-object">String</span> building = details.build.building[0]
	<span class="code-keyword">if</span>(building.equals(<span class="code-quote">"<span class="code-keyword">true</span>"</span>)) {
		<span class="code-keyword">return</span> <span class="code-keyword">true</span>
	}
	<span class="code-keyword">return</span> <span class="code-keyword">false</span>
}
</pre>
</div></div> 

<p>Not really an answer to the original bug, but might help you...</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kutzi">kutzi</a>:</i>
<p>Assigning to Kohsuke as I agree that this is a major issue which shouldn't stay open that long.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scooper4711">scooper4711</a>:</i>
<p>We're using the maven-jenkins-plugin (<a href="http://evgeny-goldin.com/wiki/Maven-jenkins-plugin" class="external-link" target="_blank" rel="nofollow noopener">http://evgeny-goldin.com/wiki/Maven-jenkins-plugin</a>) to have our build configurations under version control and to have the ability for jobs to inherit values from other jobs.<br/>
We too suffer from having to wait for all jobs to complete before we can re-load the configuration after the plugin runs. It's annoying. I'd love to see this bug fixed.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=antifun">antifun</a>:</i>
<p>I did some investigation into this, and here's what's going on:</p>

<ul>
	<li>The Reload Configuration operation from the UI attempts to load up all the jobs and runs it can find.</li>
	<li>It loads runs from disk based on the presence of a build.xml in the directory for a particular build number (cf. RunMap.load()).</li>
	<li>Runs that are in progress <b>do not</b> have a build.xml written to the directory in the course of normal operations (problem #1).</li>
</ul>


<p>Attempt 1: write a piece of code that, at the start of the reload operation (Jenkins.reload()), looks at all running Executors and forces them to marshal their state to disk (creating a build.xml). This would then allow Jenkins to notice those jobs running when it restarts.</p>

<p>Then I discovered that the unmarshal procedure for build.xml -> Run assumes that any build.xml it sees is for a run that is in State.COMPLETED. The State is not itself persisted as there's no getState() for xstream to call.</p>

<p>Attempt 2: Make the State of a run persist appropriately so that it can be recovered when Jenkins reloads the configuration.</p>

<p>This seems to work OK at least in limited tests, and I intend to put up a pull request to let people see the changes I'm proposing in the code. I do wonder, though, what the justification for doing it this way in the first place was; it seems likely that you would <em>not</em> in all cases want Jenkins to totally "trust" the state on disk when starting up, for example if there were a large time or configuration delta between the stop and start. I do, however, think that the specific case of mashing the "Reload Configuration" link should be able to assume that what was running before is still running. Things that could possibly go wrong now would mostly be in the arena of jobs that Jenkins now thinks are running but actually aren't anymore. In a reload case, you can probably expect that to happen minimally (if at all).</p>

<p>Another idea I had was based on the fact that the executors clearly are reporting back the fact that they are running a particular job, even if Jenkins doesn't believe that run actually exists due to this bug. It might be possible/better to, instead of persisting everything to disk on the Jenkins master side, have it query the slaves for running jobs when it comes back up and use the information it gets from them to reconstruct its own idea of what is currently running. I don't know how people feel about potential for abuse there, given that it would require the master to "trust" the slaves to tell it what they were working on when it restarted. A combination of the two approaches might be best (trust, but verify).</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=antifun">antifun</a>:</i>
<p>Change is here:</p>

<p><a href="https://github.com/antifun/jenkins/commit/2793ce3721de8c6a1e9d7912b92e437d6e827b2e" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/antifun/jenkins/commit/2793ce3721de8c6a1e9d7912b92e437d6e827b2e</a></p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=andrei_pozolotin">andrei_pozolotin</a>:</i>
<p>possibly related: #<a href="https://issues.jenkins.io/browse/JENKINS-14123" title="rootPOM disappears on job edit in web interface" class="issue-link" data-issue-key="JENKINS-14123"><del>JENKINS-14123</del></a></p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jfrank">jfrank</a>:</i>
<p>It is nice to see some work being done on this issue. Thank you Matt.</p>

<p>There seems to be a question of what the best approach would be, for solving the problem. Perhaps this is naive, but... When loading a new config.xml, how about doing whatever the interactive web form does? When I reconfigure a job via html form (POSTing to /job/X/configSubmit) the job is reconfigured properly and the running build list maintains existing builds just fine. In fact my workaround for this problem is to emulate the web form posting, and it has been doing great. So, could the solution here be to duplicate the functionality of /job/X/configSubmit into the code that handles /job/X/config.xml (and a similar thing for /reload if possible)? I'm not the expert though, just hoping that thought sparks some ideas.</p>

<p>Cheers.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=antifun">antifun</a>:</i>
<p>POSTing config.xml, which looks like it goes through AbstractItem.doConfigDotXml(), has similar behavior to the reload case as near as I can tell, which is to just load whatever exists in the stored state on disk (in the case of POSTing config.xml, that's whatever you just POSTed) without writing any information into the build.xml for the build numbers which might currently be in progress. So that wouldn't be affected by my change at present. I'm still trying to figure out what the other code paths are doing that is different, but it might be that the answer is "nothing in particular", and in so doing are maintaining their in-memory view of the current Jenkins object graph which contains the running builds.</p>

<p>We don't use the POST of config.xml here but I could add the same logic in to write out the build.xml (and therefore the running build state) for running builds of that job when the POST occurs, which should accomplish the same end for that usage. I'm more interested to know what people think about the general idea about writing the build.xml state out to disk for non-terminated builds. It seems like an eminently reasonable idea to me, but changing that assumption will probably require other changes. Another possibility that I haven't explored is somehow preserving the portion of the Jenkins object graph that contains the running job information, but that would require some gymnastics around saving the RunMap objects off of the Jobs before dumping them and reloading them, but then we'd have to update all the references to point the Runs at the new Job objects and so on. That feels at least as dangerous to me as persisting the state of non-terminated builds.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=antifun">antifun</a>:</i>
<p>updateByXml actually looks kind of bonkers to me anyway. It doesn't call AbstractItem.save() first, thereby ensuring other aspects of the item we might want to save prior to reloading the new config.xml will get lost. I'll have to come up with a more serious patch for the POST config.xml fix than I thought.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jfrank">jfrank</a>:</i>
<p>Any application can halt without warning, due to power out, hardware defect, OS bugs, malicious software, you-name-it, even well-intentioned code in the application itself. Therefore important state information must be persisted upon every change. Anything else risks losing that data. So by this logic, I agree that saving the state of running builds is "eminently reasonable" as you say. You have my vote <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/>.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=deepchip">deepchip</a>:</i>
<p>At the very least, do not let the administrator reload the configuration from disk without a warning saying that in-flight jobs will be lost.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>:</i>
<p>This problem was fixed in 1.473 under <a href="https://issues.jenkins.io/browse/JENKINS-12318" title="A current active build in the build history is lost if the job configuration XML uploaded" class="issue-link" data-issue-key="JENKINS-12318"><del>JENKINS-12318</del></a>.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kscaldef">kscaldef</a>:</i>
<p>We are running 1.479 and I still see this bug.  I'd like to suggest that it should be reopened.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>@vige I am not sure what build gets the hotfix but I think not 1.507. This got filed as <a href="https://issues.jenkins.io/browse/JENKINS-17337" title="Failed to load job on Matrix jobs" class="issue-link" data-issue-key="JENKINS-17337"><del>JENKINS-17337</del></a>.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dogfood">dogfood</a>:</i>
<p>Integrated in <span class="image-wrap" style=""><img src="http://ci.jenkins-ci.org/images/16x16/blue.png" style="border: 0px solid black" /></span> <a href="http://ci.jenkins-ci.org/job/jenkins_main_trunk/2411/" class="external-link" target="_blank" rel="nofollow noopener">jenkins_main_trunk #2411</a><br/>
     <span class="error">&#91;FIXED JENKINS-3265&#93;</span> Hotfix for regression in matrix project loading from 25084f5.(cherry picked from commit 3370e6edf5e0e7c7a8cf22e4a7a5decc24c45205) (Revision 6b1f8d1aaa2179c99af33a8d60851abd032442ac)</p>

<p>     Result = SUCCESS<br/>
kohsuke : <a href="https://github.com/jenkinsci/jenkins/commit/6b1f8d1aaa2179c99af33a8d60851abd032442ac" class="external-link" target="_blank" rel="nofollow noopener">6b1f8d1aaa2179c99af33a8d60851abd032442ac</a><br/>
Files : </p>
<ul>
	<li>core/src/main/java/hudson/matrix/MatrixProject.java</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jmiller">jmiller</a>:</i>
<p>We are still seeing this issue on 1.518 but only occasionally. Is anyone else seeing it?</p>

<p>Unfortunately we have not identified any particular characteristics of how or when this happens or if there is anything of note about the jobs it is happening with. If anyone has suggestions for things to look for to help diagnose the problem that would be most helpful.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dogfood">dogfood</a>:</i>
<p>Integrated in <span class="image-wrap" style=""><img src="http://ci.jenkins-ci.org/images/16x16/blue.png" style="border: 0px solid black" /></span> <a href="http://ci.jenkins-ci.org/job/jenkins_main_trunk/2710/" class="external-link" target="_blank" rel="nofollow noopener">jenkins_main_trunk #2710</a><br/>
     <a href="https://issues.jenkins.io/browse/JENKINS-3265" title="Reload Configuration from Disk (or POSTing config.xml) loses info on running builds" class="issue-link" data-issue-key="JENKINS-3265"><del>JENKINS-3265</del></a> <a href="https://issues.jenkins.io/browse/JENKINS-17341" title="Promoted builds throw NullPointerException after upgrade to 1.507" class="issue-link" data-issue-key="JENKINS-17341"><del>JENKINS-17341</del></a> Even if older plugins throw an NPE during reload, continue loading job. (Revision ef08900292726bb35be269ce6d41d8dc141c6dea)</p>

<p>     Result = SUCCESS<br/>
Jesse Glick : <a href="https://github.com/jenkinsci/jenkins/commit/ef08900292726bb35be269ce6d41d8dc141c6dea" class="external-link" target="_blank" rel="nofollow noopener">ef08900292726bb35be269ce6d41d8dc141c6dea</a><br/>
Files : </p>
<ul>
	<li>core/src/main/java/hudson/model/AbstractProject.java</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mpapo">mpapo</a>:</i>
<p>Hi all,</p>

<p>Exactly the same problem : reload the config from disk cause current running builds to disappear after reload but the build directory is present on the disk.</p>

<p>Environment: Jenkins 1.593 on RHEL 6.0</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danielbeck">danielbeck</a>:</i>
<p>Please keep issues as old as this resolved/closed.</p>

<p>So please file a new issue. Make sure it also happens on the current weekly release, as there've been possibly relevant changes since 1.593. Provide all the information asked for on <a href="https://wiki.jenkins-ci.org/display/JENKINS/How+to+report+an+issue" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/How+to+report+an+issue</a> and try to provide a minimal test case with clear instructions how to reproduce the problem.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=JIRAUSER143558">JIRAUSER143558</a>:</i>
<p>which path it refers to load configuration from local if we have 3 jenkins running on the same server</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally depends on: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-06-a2/issues?q=in%3Atitle%20JENKINS-17337">JENKINS-17337</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally depends on: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-06-a2/issues?q=in%3Atitle%20JENKINS-17341">JENKINS-17341</a>]</i>
